webknossos/webknossos/skeleton/entities.py

```diff
@@ -0,0 +1,479 @@
+            nml=_nml or self._nml,
+            enforce_id=_enforce_id,
+        )
+        self.children.append(new_graph)
+
+        return new_graph
+
+    def add_group(
+        self,
+        name: str,
+        children: Optional[List[GroupOrGraph]] = None,
+        _enforce_id: int = None,
+    ) -> "Group":
+
+        new_group = Group(name, children or [], nml=self._nml, enforce_id=_enforce_id)  # type: ignore
+        self.children.append(new_group)
+        return new_group
+
+    def get_total_node_count(self) -> int:
+        return sum(
+            len(synapse_graph.get_nodes()) for synapse_graph in self.flattened_graphs()
+        )
+
+    def get_max_graph_id(self) -> int:
+        # Chain with [0] since max is not defined on empty sequences
+        return max(
+            itertools.chain((graph.id for graph in self.flattened_graphs()), [0])
+        )
+
+    def get_max_node_id(self) -> int:
+        # Chain with [0] since max is not defined on empty sequences
+        return max(
+            itertools.chain(
+                (graph.get_max_node_id() for graph in self.flattened_graphs()),
+                [0],
+            )
+        )
+
+    def flattened_graphs(self) -> Generator["WkGraph", None, None]:
+        for child in self.children:
+            if isinstance(child, Group):
+                yield from child.flattened_graphs()
+            else:
+                yield child
+
+    def flattened_groups(self) -> Generator["Group", None, None]:
+        for child in self.children:
+            if isinstance(child, Group):
+                yield child
+                yield from child.flattened_groups()
+
+    def as_legacy_group(self) -> "LegacyGroup":  # type: ignore
+
+        return legacy_wknml.Group(
+            self.id,
+            self.name,
+            children=[
+                g.as_legacy_group() for g in self.children if isinstance(g, Group)
+            ],
+        )
+
+    def __hash__(self) -> int:
+        return hash((self._nml.id, self.id))
+
+
+@attr.define()
+class Node:
+    position: Vector3
+    _nml: "NML"
+    id: int = attr.ib(init=False)
+    comment: Optional[str] = None
+    radius: Optional[float] = None
+    rotation: Optional[Vector3] = None
+    inVp: Optional[int] = None
+    inMag: Optional[int] = None
+    bitDepth: Optional[int] = None
+    interpolation: Optional[bool] = None
+    time: Optional[int] = None
+
+    is_branchpoint: bool = False
+    branchpoint_time: Optional[int] = None
+    _enforce_id: Optional[int] = None
+
+    def __attrs_post_init__(self) -> None:
+        if self._enforce_id is not None:
+            self.id = self._enforce_id
+        else:
+            self.id = self._nml.element_id_generator.__next__()
+
+        self.position = vector3_as_float(self.position)
+
+    def __hash__(self) -> int:
+        return hash((self._nml.id, self.id))
+
+    def __eq__(self, other: object) -> bool:
+        return hash(self) == hash(other)
```

Comment:
These methods are not necessary anymore, since I changed the nxgraph usage to use ids as keys. Therefore, I would  simply remove the methods and rely on the python default behavior. Do you agree?